File: C/Sorting-Methods/Bubble_sort.c
#include <stdio.h>

void BubbleSort(int A[], int n);

int main()
{
    printf("-------- BUBBLE SORT --------- \n\n");
    int n, arr[50];

    printf("Enter no. of values: ");
    scanf("%d", &n);

    printf("Enter array of values (with spaces): ");
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &arr[i]);
    }

    BubbleSort(arr, n);

    printf("\nValues after Sorting: ");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }

    printf("\n");
    return 0;
}

void swap(int *p, int *q)
{
    int temp;

    temp = *p;
    *p = *q;
    *q = temp;
}

void BubbleSort(int A[], int n)
{
    int flag;

    // for passes
    for (int i = 0; i < n - 1; i++)
    {
        flag = 0;

        // heaviest element is sorted (thereby reducing no. of comparison)
        for (int j = 0; j < n - 1 - i; j++)
        {
            if (A[j] > A[j + 1])
            {
                swap(&A[j], &A[j + 1]);
                flag = 1;
            }
        }

        // makes algorithm adaptive - dont check for comparison if array is already sorted
        if (flag == 0)
            break;
    }
}

File: Javascript/Sorting/BubbleSort.js
const swap1 = function (a, b) {
  return ([a, b] = [b, a])
}

const swap2 = function (a, b) {
  let temp = a
  a = b
  b = temp

  return [a, b]
}

const swap3 = function (a, b) {
  a = a + b
  b = a - b
  a = a - b

  return [a, b]
}

const swap4 = function (a, b) {
  a = a ^ b
  b = a ^ b
  a = a ^ b

  return [a, b]
}

function BubbleSort(arr = []) {
  let flag

  /* number of passes */
  for (let i = 0; i < arr.length; i++) {
    flag = 0

    /* comparison and swaps */
    for (let j = 0; j < arr.length - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        ;[arr[j], arr[j + 1]] = swap1(arr[j], arr[j + 1])
        flag = 1
      }
    }

    /*  makes algorithm adaptive - dont check for comparison if array is already sorted */
    if (!flag) return arr
  }

  return arr
}

const arr = [129, 28, 10, 90, 56, 19, 100, 49, -2, 0]

console.log(BubbleSort(arr))
